### PR TITLE
FIX(client): Issues when updating installed plugins

### PR DESCRIPTION
**Pull request by Krzmbrzl**
The first issue was that the old plugin would simply get overwritten
without at least warning the user that such a thing will take place.
With this commit this is still not guaranteed as plugins can be
installed in different directories in which case no warning is emitted
(the user will end up with two versions of the plugin installed in
parallel).

The second issue was that if a plugin was loaded while it is being
updated, on some OS (e.g. Windows) the corresponding library file is
locked and can thus not be overwritten. Therefore plugins are now
explicitly cleared before an overwrite is attempted.

### Checks

- [x] My commits follow the [commit guidelines](https://github.com/mumble-voip/mumble/blob/master/COMMIT_GUIDELINES.md)

### CODE REVIEW

File: src/mumble/PluginInstaller.cpp [hunk 1]
```diff
@@ -161,18 +163,56 @@ void PluginInstaller::init() {
...
 	if (m_pluginSource == m_pluginDestination) {
 		// Apparently the plugin is already installed
-		return;
+		return false;
 	}
 
 	if (m_pluginDestination.exists()) {
+		// This most likely means that we already have some version of this plugin installed. Figure out which
+		// one that is and ask the user for confirmation for overwriting it.
+		// NOTE: Because we currently have multiple install directories this branch may not execute if the old
```
Comment (davidebeatrici) [inline src/mumble/PluginInstaller.cpp:180]:
```suggestion
		// NOTE: Because we currently have multiple install directories, this branch may not execute if the old
```

File: src/mumble/PluginInstaller.cpp [hunk 2]
```diff
@@ -161,18 +163,56 @@ void PluginInstaller::init() {
...
+					.arg(oldPlugin->getName())
+					.arg(QString(oldPlugin->getVersion())));
+
+			if (result != QMessageBox::StandardButton::Yes) {
+				// Abort as the user did not specify that they want to proceed
+				return false;
+			}
+
+			// If we proceed we have to make sure the plugin is unloaded as otherwise Mumble
+			// could still hold a handle on the underlying library file which on some OS (e.g. Windows)
```
Comment (davidebeatrici) [inline src/mumble/PluginInstaller.cpp:207]:
```suggestion
			// could still hold a handle to the underlying library file, which on some OS (e.g. Windows)
```

**Review by davidebeatrici — APPROVED** [review]

**Comment by hbeni** [comment]
Is there some nightly Installer I can download for testing?

**Comment by davidebeatrici** [comment]
Yes, you can download it from Azure Pipelines.

**Comment by hbeni** [comment]
Yes but which one? The latest i see is from friday - is the code already there (i assume no, since its from master branch?)

**Comment by davidebeatrici** [comment]
https://dev.azure.com/Mumble-VoIP/Mumble/_build/results?buildId=4987

**Comment by hbeni** [comment]
Ok did a test.
In the minute reported as 20:32 i tried three times to install the plugin.
it looks like the shutdown is not initiated, at least nothing in that direction is logged.
[fgcom-mumble.log](https://github.com/mumble-voip/mumble/files/6722269/fgcom-mumble.log)

The line `FGCom [2021-06-27 20:33:19.441]: Shutdown plugin` was when I finally shutdown the client.

**Comment by Krzmbrzl** [comment]
That's weird. The shutdown message of your plugin was always printed for me but the file still kept being locked 👀

**Comment by hbeni** [comment]
Really, weird.
Did you really see the shutdown msg from the unload, or was it from mumble-closing like i observe?


I really have no clue what could cause dll locking, but i will try to search the interwebs for it.

**Comment by davidebeatrici** [comment]
https://github.com/mumble-voip/mumble/blob/cc037569471c4bf614a84b7d1f7cef20edefa235/src/mumble/PluginManager.cpp#L549-L562

https://github.com/mumble-voip/mumble/blob/baeddddec4c301389a088a9f7059a4b5d28c7dd1/src/mumble/Plugin.h#L108

@Krzmbrzl Is perhaps the shared pointer being used somewhere?

**Comment by Krzmbrzl** [comment]
> Did you really see the shutdown msg from the unload, or was it from mumble-closing like i observe?

I did observe it when installing the plugin because I saw it before closing Mumble.

> Is perhaps the shared pointer being used somewhere?

That was what I thought at first as well but
1. It worked as expected with a different plugin
2. I verified the plugin's destructor to be called and that Qt claims to have successfully unloaded the underlying library. On that note though, I did not explicitly verify that it was the fgcom plugin (but then again: it would be odd if it was a different one) :thinking:

**Comment by hbeni** [comment]
Whats different at your test plugin?

**Comment by Krzmbrzl** [comment]
> Whats different at your test plugin?

I don't know. It was the ACRE2 plugin that I tested this with and with that it just worked as expected :shrug:

**Comment by hbeni** [comment]
At least this tells its plugin sided.
Is there source i can look for?

**Comment by Krzmbrzl** [comment]
Yes there is: https://github.com/IDI-Systems/acre2/tree/mumble-plugin/extensions/src/ACRE2Mumble

**Comment by hbeni** [comment]
Ok, i opened https://github.com/hbeni/fgcom-mumble/issues/131 for further tracking.
It would be cool if someone could help with this, i feel really alone with this dll locking stuff.

**Comment by Krzmbrzl** [comment]
Okay after a fair amount of troubleshooting it turned out that the issue was indeed on the plugin's side (more specifically an issue with OpenSSL).
Now that it has been fixed the issue seems to have been resolved.

Thus I am merging this now.

**Comment by davidebeatrici** [comment]
What kind of issue with OpenSSL?

**Comment by hbeni** [comment]
See here: https://github.com/hbeni/fgcom-mumble/issues/131#issuecomment-877094293

TLDR; the windows plugin dll got locked and not released when unloading, preventing the updater from replacing the plugin with the new version.
Cause was a change in OpenSSL which could be mitigated by adding the compile time configure option `no-pinshared`.

**Comment by davidebeatrici** [comment]
Oh, I see. For reference: microsoft/vcpkg#12056
